merge_tops : throw better error if top is invalid

zaz opened the pull request:
This is a step towards fixing [#9950](https://github.com/QubesOS/qubes-issues/issues/9950). It doesn't show the file name that caused the error, so I'll look into adding a check higher level, too.

zaz committed:
merge_tops : throw better error if top is invalid

marmarek commented:
Full json dump of the top file won't be very readable... maybe limit it somehow (first X chars?). Ideally it should point at the file and line, but that's rather hard to do at this abstraction level...

zaz commented:
@marmarek Thanks a lot! Yes, I agree. I am working on adding some nested loops in `get_tops` as that has the filepath. By the time it reaches `merge_tops` it should be correct. It might still be a good idea to keep something like the above so if all else fails the user can infer from the JSON data where the error occurred (e.g. by using `grep` for some of the output). But perhaps you have a better way of wording the error or would rather specify in the docstring that the input must be checked before being passed to `merge_tops`.

zaz commented:
@marmarek Thanks a lot! Yes, I agree. I am working on adding some nested loops in `get_tops` as that has the filepath. By the time it reaches `merge_tops` it should be correct. It might still be a good idea to keep something like the above so if all else fails the user can infer from the JSON data where the error occurred (e.g. by using `grep` for some of the output). But perhaps you have a better way of wording the error or would rather specify in the docstring that the input must be checked before being passed to `merge_tops`.

Here's the [WIP gist of what I was thinking](https://github.com/zaz/qubes-mgmt-salt-base-topd/blob/better-errors-wip/_modules/topd.py). I switched from throwing an error to logging a warning because I thought that made the output a lot cleaner, but unfortunately that caused an error to come from the minion:

```
debian:
      ----------
      _error:
          The command resulted in a non-zero exit code
      parsed:
          None
      retcode:
          126
      stderr:
          Request refused
      stdout:
  Cannot open display "default display"
  /usr/lib/python3.13/site-packages/salt/client/ssh/__init__.py:1606: DeprecationWarning: 'maxsplit' is passed as positional argument
    shim_command = re.split(r"\r?\n", stdout, 1)[0].strip()
 ```
 
I'm guessing the `merge_tops` from `topd.py` runs in dom0 and then the `merge_tops` macro from `top.jinja` runs on the minion? Did I get that right?
 
Sorry I haven't had time to fully understand how the whole system works. I don't believe it is documented anywhere?

zaz commented:
@marmarek Thanks a lot! Yes, I agree. I am working on adding some nested loops in `get_tops` as that has the filepath. By the time it reaches `merge_tops` it should be correct. It might still be a good idea to keep something like the above so if all else fails the user can infer from the JSON data where the error occurred (e.g. by using `grep` for some of the output). But perhaps you have a better way of wording the error or would rather specify in the docstring that the input must be checked before being passed to `merge_tops`.

Here's the [WIP gist of what I was thinking](https://github.com/zaz/qubes-mgmt-salt-base-topd/blob/better-errors-wip/_modules/topd.py). I switched from throwing an error to logging a warning because I thought that made the output a lot cleaner, but unfortunately that caused an error to come from the minion:

```
debian:
      ----------
      _error:
          The command resulted in a non-zero exit code
      parsed:
          None
      retcode:
          126
      stderr:
          Request refused
      stdout:
  Cannot open display "default display"
  /usr/lib/python3.13/site-packages/salt/client/ssh/__init__.py:1606: DeprecationWarning: 'maxsplit' is passed as positional argument
    shim_command = re.split(r"\r?\n", stdout, 1)[0].strip()
 ```
 
I'm guessing the `merge_tops` from `topd.py` runs in dom0 and then the `merge_tops` macro from `top.jinja` runs on the minion? Did I get that right?
 
Sorry I haven't had time to fully understand how the whole system works. I don't believe it is documented anywhere?

zaz commented:
Actually, I'm thinking instead of looping over all the tops to check them before passing to `merge_tops` it might be cleaner to pass the filenames to `merge_tops` to allow for  better error logging there.
